File: src/lib/scala/Slate.Common/src/main/scala/slate/common/subs/Subs.scala
/**
  * <slate_header>
  * author: Kishore Reddy
  * url: https://github.com/kishorereddy/scala-slate
  * copyright: 2016 Kishore Reddy
  * license: https://github.com/kishorereddy/scala-slate/blob/master/LICENSE.md
  * desc: a scala micro-framework
  * usage: Please refer to license on github for more info.
  * </slate_header>
  */
package slate.common.subs

import slate.common.{DateTime, ListMap}

class Subs(setDefaults:Boolean = true) {
  private val _groups = new ListMap[String, (Sub) => String]()

  if(setDefaults) {
    defaults()
  }

  def contains(key:String): Boolean = _groups.contains(key)


  /**
    * adds a key/value to this collection
    *
    * @param key
    * @param value
    */
  def update(key:String, value:(Sub) => String) = {
    _groups(key) = value
  }


  /**
    * gets the value with the supplied key
    *
    * @param key
    * @return
    */
  def apply(key:String):String =
  {
    if(!_groups.contains(key))
      return ""
    val sub = _groups(key)
    sub(new Sub(key, SubConstants.TypeText, -1, -1))
  }


  /**
    * parses the text into a list of substitutions
    *
    * @param text
    * @return
    */
  def parse(text:String): List[Sub] = {
    new SubParser(text).parse()
  }


  def resolve(text:String): Option[String] = {
    val tokens = parse(text)
    if(tokens == null || tokens.size == 0){
      return Some(text)
    }
    var finalText = ""
    for(token <- tokens){
      if(token.subType == SubConstants.TypeText){
        finalText += token.text
      }
      else if(token.subType == SubConstants.TypeSub){
        finalText += this(token.text)
      }
    }
    Some(finalText)
  }


  private def defaults():Unit = {

    // Default functions.
    _groups("today"     )  = ( sub ) => { DateTime.today().toStringYYYYMMDD()             }
    _groups("yesterday" )  = ( sub ) => { DateTime.today().addDays(-1).toStringYYYYMMDD() }
    _groups("tomorrow"  )  = ( sub ) => { DateTime.today().addDays(1).toStringYYYYMMDD()  }
    _groups("t"         )  = ( sub ) => { DateTime.today().toStringYYYYMMDD()             }
    _groups("t-1"       )  = ( sub ) => { DateTime.today().addDays(-1).toStringYYYYMMDD() }
    _groups("t+1"       )  = ( sub ) => { DateTime.today().addDays(1).toStringYYYYMMDD()  }
    _groups("today+1"   )  = ( sub ) => { DateTime.today().addDays(1).toStringYYYYMMDD()  }
    _groups("today-1"   )  = ( sub ) => { DateTime.today().addDays(-1).toStringYYYYMMDD() }
  }
}

File: src/common/utils/src/main/kotlin/kiit/utils/templates/Subs.kt
/**
 *  <kiit_header>
 * url: www.kiit.dev
 * git: www.github.com/slatekit/kiit
 * org: www.codehelix.co
 * author: Kishore Reddy
 * copyright: 2016 CodeHelix Solutions Inc.
 * license: refer to website and/or github
 *  </kiit_header>
 */

package kiit.utils.templates

import kiit.common.DateTimes
import kiit.common.ext.toStringYYYYMMDD
import kiit.common.values.ListMap

/**
 * Performs dynamic substitutions of variables in text.
 * Similar to interpolated strings, but at runtime. This allows for creating
 * @param items
 * @param setDefaults
 */
class Subs(items: List<Pair<String, (TemplatePart) -> String>>? = null, setDefaults: Boolean = true) {

    private val groups = ListMap(items ?: listOf())

    init {

        // Add the default subs/variables
        if (setDefaults) {
            defaults()
        }

        // Add the custom variables
        items?.let { it ->
            it.forEach { entry -> groups.add(entry.first, entry.second) }
        }
    }

    /**
     * whether this contains a substitution with the given key
     *
     * @param key
     * @return
     */
    fun contains(key: String): Boolean = groups.contains(key)

    /**
     * Size of the substitutions
     * @return
     */
    val size: Int get() = groups.size

    /**
     * gets the value with the supplied key
     *
     * @param key
     * @return
     */
    operator fun get(key: String): String = lookup(key)

    /**
     * gets the value with the supplied key
     *
     * @param key
     * @return
     */
    fun lookup(key: String): String =
            if (!groups.contains(key)) {
                ""
            } else {
                val sub = groups[key]
                sub?.invoke(TemplatePart(key, TemplateConstants.TypeText, -1, -1)) ?: ""
            }

    private fun defaults() {

        // Default functions.
        groups.add("today"    , {  DateTimes.today().toStringYYYYMMDD() })
        groups.add("yesterday", {  DateTimes.today().plusDays(-1).toStringYYYYMMDD() })
        groups.add("tomorrow" , {  DateTimes.today().plusDays(1).toStringYYYYMMDD() })
        groups.add("t"        , {  DateTimes.today().toStringYYYYMMDD() })
        groups.add("t-1"      , {  DateTimes.today().plusDays(-1).toStringYYYYMMDD() })
        groups.add("t+1"      , {  DateTimes.today().plusDays(1).toStringYYYYMMDD() })
        groups.add("today+1"  , {  DateTimes.today().plusDays(1).toStringYYYYMMDD() })
        groups.add("today-1"  , {  DateTimes.today().plusDays(-1).toStringYYYYMMDD() })
    }
}
